src/radix_connect/app_link_parser.rs

```diff
@@ -0,0 +1,154 @@
+use crate::prelude::*;
+use url::form_urlencoded;
+use url::Url;
+
+use super::{DappRequest, LinkRequest, MobileConnectRequest};
+
+#[uniffi::export]
+pub fn parse_request_url(url: String) -> Result<MobileConnectRequest> {
+    let connect_url = Url::parse(CONNECT_URL).unwrap();
+    match Url::parse(url.as_str()) {
+        Ok(parsed_url) => {
+            if parsed_url.host_str() != connect_url.host_str()
+                || parsed_url.scheme() != connect_url.scheme()
+            {
+                return Err(CommonError::RadixConnectMobileInvalidRequestUrl);
+            }
+
+            let query_parameters = parsed_url
+                .query_pairs()
+                .into_owned()
+                .collect::<HashMap<String, String>>();
+            if let Some(session_id) =
+                query_parameters.get(CONNECT_URL_PARAM_SESSION_ID)
+            {
+                match query_parameters.get(CONNECT_URL_PARAM_ORIGIN) {
+                    Some(origin) => {
+                        if let Ok(url) = Url::parse(origin) {
```

Comment:
you could probably do something like:
```rust
Url::parse(origin)
   .map_error(|_| CommonError::RadixConnectMobileInvalidOrigin)
   .and_then(|url| {
      Ok(MobileConnectRequest::Link(
                                LinkRequest {
                                    origin: url,
                                    session_id: SessionID(
                                        session_id.to_owned(),
                                    ),
                                },
                            ));
   })
```
I don't recall the exact syntax, but something along these lines :)

Reply:
you can use `map` instead of `and_then`, removing need for `Ok`.

@jakub-rdx also call `cargo clippy --fix --allow-staged`, it can fix some stuff automatically...